### PR TITLE
Optimization for slow AsyncGauge execution

### PR DESCRIPTION
**Pull request by gaojieliu**
This PR introduces a dynamic way to track slow AsyncGauge metric execution and tries not to block the caller thread as much as possible. In the high-level, this PR introduces a `AsyncGaugeExecutor`, which implements the following strategy:
 1. There are two executors and one for regular metrics and the other one is for slow metrics.
 2. All the metric evaluations are triggered by the caller.
 3. If the actual metric execution time exceeds the configured slow metric threshold, it will be moved to slow metric tracking map, which indicates the following behaviors: a. The next metric measurement call will return the cached value immediately. b. The submitted measurable will be executed asynchronously. c. If the actual measurement runtime latency becomes lower than the slow metric threshold, it will be moved out of slow metric tracking map.
 4. If the actual metric execution time belows the configured slow metric threshold, the following behaviors will be observed: a. After submitting the measurable to the regular executor, it will wait up to configured {@link AsyncGaugeExecutor#initialMetricsMeasurementTimeoutInMs} to collect the latest result. b. If it can't collect the latest value in step #a, the next call will examine the previous execution to decide whether it should be put into the slow metric tracking map or not.

 5. There is an async thread to clean up inactive metrics from slow metric tracking map to avoid the accumulation of garbage because of metric deletion.

There are several config params of `AsyncGaugeExecutor` and the user can tune it according to the actual load pattern, and the caller can construct a global `AsyncGaugeExecutor` and pass it to `MetricsRepository` via `MetricConfig`.

### CODE REVIEW

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java [hunk 1]
```diff
@@ -60,89 +68,333 @@ public void record(double value, long now) {
...
         }
-        // Always try to get the freshest measurement value
-        // Reason: let's say the initial wait time is 500ms, and the previous measurement took 600ms. In this case, if we
-        //         return the previous measurement value, it would be 59 seconds stale, assuming the measurement interval is 1 minute.
-        return submitNewMeasurementTask(config, now, asyncGaugeConfig);
+      };
+    }
+
+    public synchronized double measure(AsyncGauge asyncGauge, MetricConfig config, long now) {
+      boolean isSlowMetric = slowAsyncGaugeAccessMap.containsKey(asyncGauge);
```
Comment (huangminchn) [inline src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:262]: Do we need to override the "equals" function of AsyncGauge to make it work as a map key?

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java [hunk 2]
```diff
@@ -60,89 +68,333 @@ public void record(double value, long now) {
...
+                LOGGER.warn(String.format("The measurement for metric %s took %d ms; the metric value is %f, will move this metric to slow metric tracking",
+                    asyncGauge.metricName, asyncGauge.lastMeasurementLatencyInMs, asyncGauge.cachedMeasurement));
+                isSlowMetric = true;
+                slowAsyncGaugeAccessMap.put(asyncGauge, time.milliseconds());
+              }
+            } else if (isSlowMetric) {
+              slowAsyncGaugeAccessMap.remove(asyncGauge);
+              isSlowMetric = false;
+              LOGGER.info("The measurement for metric: " + asyncGauge.metricName + " took " +
+                  asyncGauge.lastMeasurementLatencyInMs + " ms, will move this metric out of slow metric tracking");
```
Comment (huangminchn) [inline src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:292]: Nitpick: s/will move/moved?

**Review by huangminchn — COMMENTED** [review]
Thanks Gaojie! Looks good overall; left some comments.

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java [hunk 3]
```diff
@@ -60,89 +68,333 @@ public void record(double value, long now) {
...
+   *       whether it should be put into the slow metric tracking map or not.
+   *
+   * 5. There is an async thread to clean up inactive metrics from slow metric tracking map to avoid the accumulation of garbage
+   *    because of metric deletion.
+   */
+  public static class AsyncGaugeExecutor implements Closeable {
+
+    public static class Builder {
+      private int metricMeasurementThreadCount = 3;
+      private int slowMetricMeasurementThreadCount = 7;
```
[the remark this answers is not in the document]
Comment (gaojieliu) [reply src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:101]: Here is the context:
There were a few metrics, which were triggered by topic deletion, while the offset measurement still kicks in, and these slow metrics occupy the thread pool for quite long time, since the total number of threads are fixed, the fast metrics didn't get a chance to execute, but just stuck in the queue of the thread pool.
Another reason of the problem is the initial wait time after submitting the measurable to the thread pool, because of the above behavior, almost every metric needs to wait up to 500ms, which caused metric gap since every collection take 10-30 mins.

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java (hunk 1 above)
Comment (gaojieliu) [reply src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:262]: I think it should work since the map will use the default implementation of `equal` and `hashCode`, and I suppose the `AsyncGauge` instance for the same metric won't change during the lifetime of the metric.
Let me if you think my understanding is not accurate.

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java (hunk 2 above)
Comment (gaojieliu) [reply src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:292]: Sure

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java [hunk 4]
```diff
@@ -60,89 +68,333 @@ public void record(double value, long now) {
...
-          if (!REDUNDANT_LOG_FILTER.isRedundantLog(warningMessagePrefix)) {
-            LOGGER.warn(String.format("%s Return the error code: %f", warningMessagePrefix, cachedMeasurement));
+          // If the last measurement future is still running but hasn't exceeded the max timeout, keep it running and return the cached value.
+          // Otherwise, cancel the last measurement future to prevent OutOfMemory issue, and submit a new measurement task.
+          if (asyncGauge.lastMeasurementStartTimeInMs < 0 // Measurable hasn't been executed yet
+              || time.milliseconds() - asyncGauge.lastMeasurementStartTimeInMs < maxMetricsMeasurementTimeoutInMs) {
+            return asyncGauge.cachedMeasurement;
+          } else {
+            asyncGauge.cachedMeasurement = maxTimeoutErrorCode;
+            asyncGauge.lastMeasurementFuture.cancel(true);
```
Comment (sushantmane) [inline src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:317]: If expectation about this `lastMeasurementFuture.cancel(true)` is to interrupt the thread then it won't work as CFs don't have underlying mapping with threads as such. 

To achieve that we'll have to rely on `FutureTask` return from the executors

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java [hunk 5]
```diff
@@ -60,89 +68,333 @@ public void record(double value, long now) {
...
-        // Reason: let's say the initial wait time is 500ms, and the previous measurement took 600ms. In this case, if we
-        //         return the previous measurement value, it would be 59 seconds stale, assuming the measurement interval is 1 minute.
-        return submitNewMeasurementTask(config, now, asyncGaugeConfig);
+      };
+    }
+
+    public synchronized double measure(AsyncGauge asyncGauge, MetricConfig config, long now) {
+      boolean isSlowMetric = slowAsyncGaugeAccessMap.containsKey(asyncGauge);
+
+      if (isSlowMetric && slowMetricsMeasurementExecutor.isShutdown()
```
Comment (sushantmane) [inline src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:264]: Can we simplify this method? It is unnecessarily complex with too much nesting...

Comment (gaojieliu) [reply src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:264]: You meant this particular statement or the whole function?

Comment (gaojieliu) [reply src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:264]: And which part is the `unnecessarily complex`?

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java (hunk 4 above)
Comment (gaojieliu) [reply src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:317]: Let me try to do some test.

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java (hunk 5 above)
Comment (sushantmane) [reply src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:264]: whole method..

There are several if-else with nesting.. it makes code to hard to follow IMHO

why not just do
```
if (cond) {
 return x;
}

// next code without else
...
```

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java (hunk 4 above)
Comment (gaojieliu) [reply src/main/java/io/tehuti/metrics/stats/AsyncGauge.java]: Good catch and I updated it and added test to validate this code path.

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java (hunk 5 above)
Comment (gaojieliu) [reply src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:264]: Done.

File: src/main/java/io/tehuti/metrics/stats/AsyncGauge.java [hunk 6]
```diff
@@ -269,67 +269,64 @@ public synchronized double measure(AsyncGauge asyncGauge, MetricConfig config, l
...
+        // Always try to get the freshest measurement value
+        // Reason: let's say the initial wait time is 500ms, and the previous measurement took 600ms. In this case, if we
+        //         return the previous measurement value, it would be 59 seconds stale, assuming the measurement interval is 1 minute.
+        return submitNewMeasurementTask(asyncGauge, config, now, isSlowMetric);
+      }
+      // If the last measurement future is still running but hasn't exceeded the max timeout, keep it running and return the cached value.
+      // Otherwise, cancel the last measurement future to prevent OutOfMemory issue, and submit a new measurement task.
+      if (asyncGauge.lastMeasurementStartTimeInMs < 0 // Measurable hasn't been executed yet
+          || time.milliseconds() - asyncGauge.lastMeasurementStartTimeInMs < maxMetricsMeasurementTimeoutInMs) {
+        return asyncGauge.cachedMeasurement;
```
Comment (huangminchn) [inline src/main/java/io/tehuti/metrics/stats/AsyncGauge.java:314]: Just to ensure this edge case doesn't exist: if the first measurement for this metric is stuck, it will never have a chance to reach the below codes?

**Review by huangminchn — APPROVED** [review]
Thanks a lot Gaojie!
